### PR TITLE
DAOS-6709 rsvc: xfer only attr size and fix tests

### PR DESCRIPTION
**Pull request by bfaccini**
Before this patch, when fetching pool/container attributes,
if a segment buffer size is greater than attribute's real size,
extra data is also leaked.
This patch fixes this, by sending each per-attr segment as a
single bulk transfer to avoid Mercury inter-segment
management and buffer packing.
Doing so preserves original/expected behaviour with no
more extra data leak, so associated tests do not need
to be modified.

Change-Id: I989c384fdc667d12d8aeba3f67e51bb8be6ab290
Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

### CODE REVIEW

**Comment by bfaccini** [comment]
4 tests failures in Functional_Hardware_Large and all unrelated to this PR :-( :

1)Test / Functional_Hardware_Large / 1-./erasurecode/ec_ior_smoke.py:ErasureCodeIor.test_ec;container-hosts-client_processes-iorflags-objectclass-1M_128M-pool-0-1-325e – FTEST_erasurecode.ErasureCodeIor
—>> DAOS-6790/DAOS-6975

2)Test / Functional_Hardware_Large / 1-./pool/create_capacity_test.py:PoolCreateTests.test_create_pool_quantity;hosts-pool-0-timeouts-6018 – FTEST_pool.PoolCreateTests
—> DAOS-6976

3)Test / Functional_Hardware_Large / 1-./pool/create_test.py:PoolCreateTests.test_create_no_space;hosts-pool-0-timeouts-474e – FTEST_pool.PoolCreateTests
—>> DAOS-6932

4)Test / Functional_Hardware_Large / 2-./pool/create_test.py:PoolCreateTests.test_create_no_space_loop;hosts-pool-0-timeouts-474e – FTEST_pool.PoolCreateTests
—>> DAOS-6932

**Comment by bfaccini** [comment]
Just would to raise to the reviewers attention what was already specified in the commit message, but this PR-4745 is a new attempt (after [PR-4623](https://github.com/daos-stack/daos/pull/4623) which had finally to be reverted due to regressions found by pool/container attribute-related subtests during "daos_test [-p,-c]") to fix issue and make "daos_test [-p,-c]" aware of new behaviour (only real attribute's value size is xfered with no extra data to fill bigger buffer sized, meaning that bulk xfer can pack more attributes value within same segment buffer if possible, and also that exact xfered size for each attributes is returned instead of exact attributes size).
This may break current definition of daos_cont_get_attr() behaviour, but I have not been able to find it clearly described in the documentation...

**Comment by johannlombardi** [comment]
@liw I would appreciate your feedback on this change.

File: src/rsvc/srv_common.c
Comment (liuxuezhao) [inline src/rsvc/srv_common.c:257]: I see you only want to transfer the real size for each attribute, but the handling here is incorrect.
1. I think you need to change attr_bulk_transfer() - start one crt_bulk_transfer() for each attribute's actual size, and set correct offset/length for each attribute.
2. the sizes setting and bulk_size setting above looks not correct to me.

You may refer obj_bulk_transfer()'s code, just FYI.

Comment (liw) [reply src/rsvc/srv_common.c:257]: Hmm, Bruno's change looks right to me. The problem is that the original author fell into the crt pitfall that crt looks at iov_buf_len instead of iov_len when creating a bulk handle, no?

**Review by liw — APPROVED** [review]

File: src/rsvc/srv_common.c
Comment (liuxuezhao) [reply src/rsvc/srv_common.c:257]: in client's attr_bulk_create(), the bulk handle is created by -
2267 >       /* values */                                                            
2268 >       if (sizes != NULL && values != NULL) {                                  
2269 >       >       for (j = 0; j < n; ++j)                                         
2270 >       >       >       if (sizes[j] > 0)                                       
2271 >       >       >       >       d_iov_set(&sgl.sg_iovs[i++],                    
2272 >       >       >       >       >         values[j], sizes[j]);                 
2273 >       }          

i.e. for the attr values, each bulk segment is set to user provided buffer list.

Then at server-side, need to start multi-bulk-transferring to fill to each attr's value's segment, if the real size of some middle segments is less than user provided buffer.
For example, client read 2 attrs, each provide 1KB buffer. But actually at server side, it finds that each attr only with value of 0.5KB. Then should start 2 bulk transfer, 1st transfer to first segment's 0-0.5KB, 2nd transfer to 2nd segment's 0-0.5KB. This is what this patch want to do, no?
or I may mis-understood it...

Comment (liw) [reply src/rsvc/srv_common.c:257]: You're right. I mistakenly assumed that the client-side buffer is one continuous segment.

Comment (bfaccini) [reply src/rsvc/srv_common.c:257]: Thanks a lot @liuxuezhao and @liw for these very interesting and helpful comments. I think I had to have a look to the similar behaviour in obj_bulk_transfer() before !!... And also it seems to me that attr_bulk_transfer() has been designed to work with a single/common buffer at some point of time...
Will try to push a new commit soon, based on your comments, many thanks again !

Comment (jolivier23) [reply src/rsvc/srv_common.c:257]: So you are planning to update this PR @bfaccini ?   I noticed that @liw approved it after that comment.

Comment (jolivier23) [reply src/rsvc/srv_common.c:257]: Or perhaps it's just github being weird.  It shows this comment block and afterward, the approval but the approval appears tied to the prior comment from @liw, not the most recent one.

**Review by jolivier23 — CHANGES_REQUESTED** [review]
Just requesting changes to make sure I don't land this til resolved.

File: src/rsvc/srv_common.c
Comment (bfaccini) [reply src/rsvc/srv_common.c:257]: > So you are planning to update this PR @bfaccini ?

@jolivier23 , yes I will try to implement the fix for DAOS-6709  based on these comments.

I think @liw had first approved and then felt ok with @liuxuezhao about the way that should be used to fix...

File: src/rsvc/srv_common.c
Comment (jolivier23) [reply src/rsvc/srv_common.c:274]: if you need to refresh, fix this one.   Not a huge issue but indent does look a bit off.

**Review by jolivier23 — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Review by liw — APPROVED** [review]

**Pull request merged**
